### PR TITLE
Configfs - DON'T MERGE

### PR DESCRIPTION
**Pull request by fhunleth**
I made a PR so that we can track our comments on this, but it's a work-in-progress, so it's not meant to be merged.

### CODE REVIEW

File: lib/nerves_runtime/configfs.ex [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  def init([]) do
+    # configfs_home="/sys/kernel/config"
+    # mount none $configfs_home -t configfs
+    case Runtime.cmd("mount", [], :return) do
+      {str, 0} -> String.contains?(str, "none on /sys/kernel/config type configfs")
+      {err, _} -> raise "Failed to check mounted partitions: #{inspect(err)}"
+    end
+    |> unless do
+      Logger.info("Going to mount ConfigFS.")
+      {_, 0} = Runtime.cmd("mount", ["none", "/sys/kernel/config", "-t", "configfs"], :return)
```
Comment (fhunleth) [inline lib/nerves_runtime/configfs.ex]: You can mount special filesystems like this from erlinit.config. Why don't you PR nerves_system_bbb and rpi0 with those mounts.

File: lib/nerves_runtime/configfs.ex [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  defp mkdir_p(dir) do
+    {_, 0} = Runtime.cmd("mkdir", ["-p", dir], :return)
+  end
+
+  defp ln_s(patha, pathb) do
+    {_, 0} = Runtime.cmd("ln", ["-s", patha, pathb], :return)
+  end
+
+  defp write(path, value) do
+    {_, 0} = Runtime.cmd("sh", ["-c", "echo #{value} > #{path}"], :return)
```
Comment (fhunleth) [inline lib/nerves_runtime/configfs.ex]: `File.write!` doesn't work on these files? Could you test? If not, OTP 21 can't come fast enough.

File: lib/nerves_runtime/configfs.ex [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      write(Path.join("/sys/kernel/config/", file), value)
+    end)
+  end
+
+  defp apply_link(patha, pathb) do
+    ln_s(patha, pathb)
+  end
+
+  defp mkdir_p(dir) do
+    {_, 0} = Runtime.cmd("mkdir", ["-p", dir], :return)
```
Comment (fhunleth) [inline lib/nerves_runtime/configfs.ex]: `File.mkdir_p!`?

File: lib/nerves_runtime/configfs.ex [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  end
+
+  defp ln_s(patha, pathb) do
+    {_, 0} = Runtime.cmd("ln", ["-s", patha, pathb], :return)
```
Comment (fhunleth) [inline lib/nerves_runtime/configfs.ex]: `File.ln_s!`?

File: lib/nerves_runtime/configfs.ex [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                }
+              }
+            }
+          },
+          "strings" => %{
+            "0x409" => %{
+              "manufacturer" => Nerves.Runtime.KV.get_active("nerves_fw_author"),
+              "product" => Nerves.Runtime.KV.get_active("nerves_fw_product"),
+              # This is obviously not a good idea.
+              "serialnumber" => Nerves.Runtime.cmd("cat", ["/proc/cpuinfo"], :return) |> elem(0) |> String.split("Serial\t\t: ") |> List.last() |> String.trim()
```
Comment (fhunleth) [inline lib/nerves_runtime/configfs.ex]: This won't work on many of our systems. Maybe that's why you said that it isn't a good idea. Calling `boardid` can get you a better serial number, but it's not perfect either.

File: lib/nerves_runtime/configfs.ex (hunk 1 above)
Comment (ConnorRigby) [reply lib/nerves_runtime/configfs.ex]: I figured as much. I guess we can assume configfs is mounted by Erlinit.

**Comment by ConnorRigby** [comment]
I'm going to address all the `File` module function calls here.

I tried using all of `File.ln_s!`, `File.write!` and `File.mkdir_p!` and they didn't seem to work properly. I will try it again next pass i take here.

**Comment by ConnorRigby** [comment]
Also i think having this code in it's own project is a fine idea also. We could even add it at the `nerves_system_rpi0` level maybe, so users won't have to add it themselves? FWIW `configfs` is not _just_ for gadget config, but that being said i don't know anything else that uses it off the top of my head.

**Comment by fhunleth** [comment]
I was afraid that the `File.*` functions didn't work on the configfs filesystem. That's annoying.

**Comment by fhunleth** [comment]
Regarding the separate project part, I think there are two options:
1. A USB gadget ethernet/serial project - that seemed like what what was being started here
2. A `configfs` project and a second project that uses it.
This code didn't look like it was intended to be generic so I was assuming the first. That's fine with me, too.

**Comment by fhunleth** [comment]
See https://github.com/nerves-project/usb_gadget for work on this.

**Pull request closed without merging**